src/lzw.ml

```diff
@@ -11,58 +20,75 @@ let rec get_bits bytedata offset count =
       let bit = (byte lsr (7 - bitoffset)) land 1 in
       (bit lsl (n - 1)) + get_bits bytedata (offset + 1) (n - 1)
 
-let build_table_entry a b =
-  match b with [] -> a | x :: _ -> List.rev (x :: List.rev a)
+let build_table_entry ~previous next =
+  if String.(equal next empty) then previous
+  else previous ^ String.make 1 next.[0]
 
 let update_buffer buffer offset codestr =
-  List.fold_left
-    (fun acc c ->
-      Cstruct.set_char buffer acc c;
-      acc + 1)
-    offset codestr
+  let len = String.length codestr in
+  Unsafe.blit_string codestr 0 buffer offset len;
+  offset + len
+
+let max_dict_size = 4096
+
+let init_table () =
+  Array.init 4096 (function
+    | i when i < 256 -> String.make 1 (Char.chr i)
+    | _ -> String.empty)
+
+let reset_table dict =
+  Array.iteri
+    (fun i _ ->
+      Array.set dict i
+        (if i < 256 then String.make 1 (Char.chr i) else String.empty))
+    dict
 
 let decode input output_buffer =
-  let dict : lzwstring array = Array.make 4096 [] in
+  let dict : lzwstring array = init_table () in
 
-  let rec inner in_offset out_offset code_size next_code_index prev_code =
+  let rec inner in_offset out_offset code_size next_code_index
+      (prev_code : lzwstring) =
     let code = get_bits input in_offset code_size in
     match code with
     | 257 -> () (* end code *)
     | 256 -> (
         (* clear code *)
+        reset_table dict;
         let next_code = get_bits input (in_offset + code_size) 9 in
         match next_code with
         | 257 -> ()
         | _ ->
-            let out_offset =
-              update_buffer output_buffer out_offset
-                [ Char.chr (next_code land 0xFF) ]
-            in
-            inner
-              (in_offset + code_size + 9)
-              out_offset 9 258
-              [ Char.chr (next_code land 0xFF) ])
+            let next = String.make 1 (Char.chr (next_code land 0xFF)) in
+            let out_offset = update_buffer output_buffer out_offset next in
+            inner (in_offset + code_size + 9) out_offset 9 258 next)
     | _ ->
         let entry =
-          if code < 256 then [ Char.chr (code land 0xFF) ]
```

Comment:
Some of these changes were me trying to understand the code better, happy to revert them

Reply:
Just depends if you think that 

```
String.make 1 (Char.chr code)
```

Is cheaper or mode expensive than an array access? I'd assume the array access is cheaper?